### ISSUE TITLE
Delete confirmation dialogue is unreadable

### ISSUE THREAD

**Issue by alexanderadam** [MrCarroll/joplin-desktop-snap#3]
When I try to delete a note, it shows only squares.
The official AppImage does not have this problem, so it's likely that it's some issue with the Snap image.

![Joplin deletion text in squares](https://user-images.githubusercontent.com/372620/91615850-c64b4d00-e984-11ea-9e94-d0385e3b7fa7.png)

**PS:** Thank you for your Snap image :+1:

**Comment by JGCarroll** [comment]
This is a big problem that a lot of snaps have. I can't say I completely understand the nuances behind it, but it's generally speaking an incompatibility between the version of fontconfig in the snap and the fontconfig on the system host. Generally I see this pop up in snaps run mostly on Arch/Manjaro and occasionally Fedora.

From what I know of the issue, this is very likely to fix it for you:
`rm -rf ~/.cache/fontconfig`
`sudo fc-cache -r -v`
`rm ~/snap/joplin-james-carroll/common/.cache/fontconfig/`


This should regenerate the system font cache, and then trigger the snap into copying it over fresh when it's next run.

As far as fixing this for everyone generally goes, there's discussion about how the snapd designers would like to approach it happening, but as far as my specific involvement goes, I'm lead to believe it'll just happen eventually when the gnome-platform snap this is dependent on incorporates the proper fix, but I couldn't say when that might happen.

**Comment by alexanderadam** [comment]
Indeed, this fixed it! Thank you! :pray: 

> Generally I see this pop up in snaps run mostly on Arch/Manjaro and occasionally Fedora.

In my case this happened on Ubuntu 20.04.

>  there's discussion about how the snapd designers would like to approach it happening

Can you please provide a link where this discussion is happening?

Thank you in advance!

**Comment by JGCarroll** [comment]
The information is spread out over a lot of threads on the Snapcraft forums and possibly elsewhere, but the single biggest thread which has the comments about how a future update to the gnome platform snaps will hopefully fix this is here:

https://forum.snapcraft.io/t/snapped-app-not-loading-fonts-on-fedora-and-arch/12484

Thanks for all the feedback, it's good to know that the font config fix does actually work; I've never personally had the problem and my understanding is 99% of people seemingly don't but there's still a substantial amount of people who have this problems with snaps and squares for letters consistently so hopefully the proposed fixes will solve this properly.

I'd expect you to not need to rerun the font cache generation again, but for any snaps you might already have installed that might have internal copies of the older cache you'll probably find removing their internal font config caches should clear them up too and hopefully any new snaps will just be fine now.

**Comment by JGCarroll** [comment]
This is fixed properly for snaps built on Snapcraft 4.4+ using a Snapcraft extension. For this Joplin Snap, versions 1.3+.

**Comment by alexanderadam** [comment]
> This is fixed properly for snaps built on Snapcraft 4.4+ using a Snapcraft extension. For this Joplin Snap, versions 1.3+.

Since this issue still occurs for me: do I have to install some extension or so to have it fixed?

**Comment by JGCarroll** [comment]
A Snapcraft extension refers to adding a single line to the snapcraft.yaml like this,

https://github.com/MrCarroll/joplin-snap/blob/163b1ca52323661dd78959d240856ba8bad61843/snap/snapcraft.yaml#L93

What it actually does is transform the build file just before its been processed and adds sensible defaults and wrappers to the snap. Particularly with 4.4+ of Snapcraft, the Gnome extensions were refactored so that they used private font caches per snap, so the font caches are generated with the Gnome platform snaps own binaries. In theory this should have been the magic fix for you, but in practise here we are. I have certainly noticed it have benefits, but edge cases still exist like yourself.

End users don't have to do anything to benefit from this, but the snap author does (Rebuild on 4.4+, using the extension, and publish it). There's a second issue with Electron apps though, because a lot of people distribute Electron snaps with Electron Builder rather than Snapcraft, and from what I see it basically ignores modern practises (E.G, doesn't build in a fresh container, doesn't benefit from improvements in the Snapcraft tool since it avoids it and writes the manifest files manually, etc). From what I remember, Electron Builder probably doesn't use extensions at all, because it only really applies to Snapcraft, the build tool, not snap/snapd. So without taking a closer look at the Super-Productivity snap itself, I'd guess it isn't using the Snapcraft Extensions at all. (I've only just seen the linked issue now but I assume it's still relevant).

My recommendation generally is that people avoid Electron Builder for snaps. Whilst it's nice to have the one build tool for everything Electron, it really isn't that hard to get by without it, and there's big benefits in treating Electron apps the same as every other snap. If you look at this repo as an example, you generally just treat it as if it was an AppImage and just extract the contents into the right folder so they end up in the snap directly instead, it's not generally too hard, and there's Github Actions and Launchpad that can simplify automating it.

As an aside, since you have had this font error for months and you seem more willing than most to try fix it, you might want to make an account at https://forums.snapcraft.io and have a chat with some of the core snapd devs there, where you might be able to get some more direct support from people about your unique machine. Otherwise, I'm hoping there'll be an extension for the Core20 base snap soon to upgrade to, and this should be another good chance at fixing this problem for you at least in Joplin's snap specifically.

**Comment by alexanderadam** [comment]
Thank you for responding so fast and clearly, James.

> the font caches are generated with the Gnome platform snaps own binaries

This probably explains, why [regenerating the font cache](https://github.com/MrCarroll/joplin-snap/issues/3#issuecomment-683216642) doesn't work anymore?

**Comment by JGCarroll** [comment]
Before 4.4, everytime you run a snap it checked to see if the font caches existed, needed updating, and actually updated it. After 4.4 (with extensions of course), it runs once everytime a snap runs its configure hook (I.E, everytime its updated, freshly installed, or the `snap set` command is used). It also started placing the output at ~/snap/example/common/fontconfig instead of the previous location.

If you were to delete it, I'm unsure what would actually happen because nothing would recreate it until the next update, it might simply continue to work but the first time a GTK dialog box opens have a small delay. 

The relevant code for making the fontconfig in the modern system is here https://github.com/snapcore/snapcraft/blob/master/extensions/desktop/common/fonts, so you might be able to recreate it yourself if you fancy experimenting.

Edit:
Consider using `snap run --shell joplin-james-carroll.joplin` so you can enter the snaps container, where you'd be able to run bash commands with the same environment the snap itself has, it might be as simple as manually running the link above.

**Comment by alexanderadam** [comment]
I just tried to run `snap run --shell joplin-james-carroll.joplin` and `fc-cache -r -v` within that container but it didn't help.

Also the message after entering the container said `To run a command as administrator (user "root"), use "sudo <command>".` but when I tried to call `fc-cache` with prepended `sudo` I'm getting `bash: /usr/bin/sudo: Permission denied`.

It seems that Snap itself still has some issues. :wink:

**Comment by JGCarroll** [comment]
I've just reread the code, it's been a while since I'd last looked at it.

Everytime the configure hook runs, it generates a cache of the system fonts, using the font cache from the Gnome platform snap. This is stored in `/var/snap/joplin-james-carroll/common/fontconfig`, not `~/snap/joplin-james-carroll/common/fontconfig`, which did exist when I installed fresh on a VM and I've no idea why. The /var fontconfig has all the actual binary cache files, and a single config file. 

You can trigger the configure hook to run by doing `sudo snap set joplin-james-carroll null=null`, so no need to mess with the `snap run --shell` above.

There's also still `~/joplin-james-carroll/current/.config/fontconfig`, the `fonts.conf` here points to the `/var` one above. So without this in place, then the system caches won't be accessible either. 

That folder is generated every time a new revision occurs, with a seperate mechanism to the config hook. You can trigger it to be rebuilt with `rm ~/snap/joplin-james-carroll/current/.last_revision`. 

So with the above two commands, we can at least get everything back to being a consistent state.

Could you give the output of `cat ~/snap/joplin-james-carroll/current/.config/fontconfig/fonts.conf` after running the two above please (after launching Joplin once)

**Comment by alexanderadam** [comment]
This is the output of `cat ~/snap/joplin-james-carroll/current/.config/fontconfig/fonts.conf`:

```xml
<fontconfig>
  <dir>/home/username/.fonts</dir>
  <include ignore_missing="yes">conf.d</include>
  <include ignore_missing="yes">/snap/joplin-james-carroll/46/gnome-platform/etc/fonts/fonts.conf</include>
  <cachedir prefix="xdg">fontconfig</cachedir>
  <cachedir>/var/snap/joplin-james-carroll/common/fontconfig</cachedir>
</fontconfig>
```

**PS:** It was `~/snap/joplin-james-carroll/current/.last_revision` for me – with an underscore at `last_revision`
**PPS:** I did both but the dialogues are still unreadable to me

**Comment by JGCarroll** [comment]
My bad, I edited the above `.last_revision` to be correct. 

I've noticed today there's progress on the Gnome extensions for the Core20 base snap, this might take a few weeks to become properly available but it's good news because the Core20 base may have a chance of improving this with the more recent font libraries it'll allow for.

In the meantime, your fontconfig above shows you have user specific fonts in ~/.fonts , just for testing purposes it might be worth renaming ~/.fonts so it's not found, regenerate the font caches on the system, delete the snaps `.last_revision` and see if it improves. If it doesn't, feel free to put your fonts back in right place, and try simply removing `<dir>/home/username/.fonts</dir>` and `<cachedir prefix="xdg">fontconfig</cachedir>` from the fontconfig in the snap manually.

Essentially, I'm thinking there's a chance that some of the local fonts in ~/.fonts are causing issues, and it might be one bad font causes the whole thing to come tumbling down, so avoiding importing them into the snap might be helpful.

**Comment by alexanderadam** [comment]
> In the meantime, your fontconfig above shows you have user specific fonts in ~/.fonts , just for testing purposes it might be worth renaming ~/.fonts so it's not found, regenerate the font caches on the system, delete the snaps `.last_revision` and see if it improves.

I just did (including `sudo snap set joplin-james-carroll null=null` just to be sure).
The conf doesn't contain the local font directory anymore:

```xml
<fontconfig>
  <include ignore_missing="yes">conf.d</include>
  <include ignore_missing="yes">/snap/joplin-james-carroll/46/gnome-platform/etc/fonts/fonts.conf</include>
  <cachedir prefix="xdg">fontconfig</cachedir>
  <cachedir>/var/snap/joplin-james-carroll/common/fontconfig</cachedir>
</fontconfig>
```
But the about dialogue still shows block gibberish instead of a proper font.

> If it doesn't, feel free to put your fonts back in right place, and try simply removing `<dir>/home/username/.fonts</dir>` and `<cachedir prefix="xdg">fontconfig</cachedir>` from the fontconfig in the snap manually.

Removing `<cachedir prefix="xdg">fontconfig</cachedir>` helped! :tada: 

> Essentially, I'm thinking there's a chance that some of the local fonts in ~/.fonts are causing issues, and it might be one bad font causes the whole thing to come tumbling down, so avoiding importing them into the snap might be helpful.

Is there a proper way to debug 'bad fonts'?

**Comment by JGCarroll** [comment]
Woohoo! Progress at last :)!

Just to confirm, is it exclusively ` <cachedir prefix="xdg">fontconfig</cachedir>` that needs to be removed to get everything working? I.E, I could just tell people to run `sed -i '/cachedir prefix="xdg"/ d' ~/snap/joplin-james-carroll/current/.config/fontconfig/fonts.conf` as a magic one liner for people with the issue?

I'm not exactly sure how to debug this further, but I think this is worth following up with Sergio in https://github.com/MrCarroll/joplin-snap/issues/7 and he might have some ideas on how to diagnose you more individually and how to apply it to the Snap ecosystem in general.

From my side, waiting on the core20 migration is still a possible fix but I could potentially also run the sed command for users by default too, it's worth considering.

**Comment by alexanderadam** [comment]
> Just to confirm, is it exclusively ` <cachedir prefix="xdg">fontconfig</cachedir>` that needs to be removed to get everything working? I.E, I could just tell people to run `sed -i '/cachedir prefix="xdg"/ d' ~/snap/joplin-james-carroll/current/.config/fontconfig/fonts.conf` as a magic one liner for people with the issue?

It seems like it. I just 'resetted' everything and got the font bug again. I stopped Joplin, removed the `cachedir prefix` and everything was readable again.

I don't understand the relations between those but it seems indeed to work.

**Comment by JGCarroll** [comment]
Thanks for all your help here. I know this problem has persisted with you for a long time and most people would probably have abandoned snaps by now as a result of it, so you've gone above and beyond here, sincereley, thanks a lot!

There's a good chance this would probably still work for super-productivity and others even if it isn't using the extensions, it might have a bit more of an obvious delay since it won't have at least some of the caches in /var, but hopefully its something.

**Comment by alexanderadam** [comment]
> Thanks for all your help here.

No, I have to thank _you_! You are a very helpful and fast-responding maintainer. :+1:
It is a pleasure to communicate with you.

**PS:** Are you maintaining all of those Snaps in your free time?

**Comment by JGCarroll** [comment]
Thanks for the kindness :)

I do maintain all these in my free time, but fortunately it doesn't take much overall, and it gives me a hobby and something to write about on my CV's so it's all good.

**Comment by alexanderadam** [comment]
This might be a strange idea:

Can you test whether removing the `cachedir` thingy is doing any harm in your tests?
If it doesn't: would it make sense to automatically remove this line when executing Joplin until this issue is solved globally?

Something like `sed …; …/bin/JoplinApp`?

**Comment by JGCarroll** [comment]
It's exactly what I was considering, however I'm holding back hoping that Core20 with the new extensions might improve things first, since I'm hopeful it might be available soon. https://github.com/snapcore/snapcraft/pull/3407

Currently, people will likely try the snap, see its broken, and either look for and acquire a fix or avoid the snap entirely. I think this is probably better than implementing a temporary fix now, doing the Core20 transition, and then possibly breaking them after they've already made some commitment.

If/when Core20 drops, assuming it doesn't fix the problem for you, I'd make the sed default. The slowdown from the lack of font caches isn't going to be noticable, but the broken fonts certainly are.

Unfortunately I've just pushed a new revision to stable tonight so you'll have to run sed again soon. But depending how quickly 1.7 and the Core 20 extension drops, it might be the last :)

**Comment by JGCarroll** [comment]
I've pushed a short lived branch of core20, would you mind giving it a go and seeing if it works with your fonts?

You can access it with `sudo snap refresh joplin-james-carroll --channel=stable/core20-beta` and it should transition back onto the normal stable releases when I pull it from the store.

**Comment by alexanderadam** [comment]
It seems to work out of the box. :+1: 
How can I migrate back to the stable release?

**Comment by JGCarroll** [comment]
When the next revision is pushed you'll migrate back to the same channel as everybody else automatically, aside from the working fonts I wouldn't imagine it to be any different to the current main stable release.

Otherwise I think the snap revert command can help you, but it'll disable automatic updates until you re-run the next refresh manually, and would revert any modifications made to your notes, so generally I'd imagine you're fine being on core20 until I push everybody else on it too for 1.7

Glad to hear this is eventually working for you out of the box :)!
